### PR TITLE
Adds multiple arrows to streamplot

### LINKED ISSUE #17740: Multiple Arrows on Streamplots
Currently, the only option when plotting streamlines is to have one arrow in the middle of each line:

image

It would be helpful when plotting geoscience data to have multiple arrows on each line, like so:

image

### PR DESCRIPTION
**Pull request by clairefio**
Closes #17740

## PR Summary
This pull request adds a parameter "numarrows" to the streamplot function to allow the user to choose how many arrows to plot on each streamline, rather that plotting just one in the middle. This would be very helpful when plotting geoscience data. Languages such as NCL (NCAR Commandline language) have this functionality:

https://www.ncl.ucar.edu/Applications/Images/stream_9_1_lg.png

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/streamplot.py [hunk 1]
```diff
@@ -193,26 +198,35 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
+        if numarrows <= -1:
+            raise ValueError("The value of numarrows must be"
+                             "greater than -1")
 
-        if isinstance(linewidth, np.ndarray):
-            line_widths = interpgrid(linewidth, tgx, tgy)[:-1]
-            line_kw['linewidth'].extend(line_widths)
-            arrow_kw['linewidth'] = line_widths[n]
+        # Add arrows along each trajectory.
+        for x in range(1, numarrows+1):
```
Comment (jklymak) [inline lib/matplotlib/streamplot.py:210]:
```suggestion
        for x in range(0, numarrows):
```

File: lib/matplotlib/streamplot.py [hunk 2]
```diff
@@ -193,26 +198,35 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
-        if use_multicolor_lines:
-            color_values = interpgrid(color, tgx, tgy)[:-1]
-            line_colors.append(color_values)
-            arrow_kw['color'] = cmap(norm(color_values[n]))
-
-        p = patches.FancyArrowPatch(
-            arrow_tail, arrow_head, transform=transform, **arrow_kw)
-        axes.add_patch(p)
-        arrows.append(p)
+            s = np.cumsum(np.hypot(np.diff(tx), np.diff(ty)))
```
Comment (jklymak) [inline lib/matplotlib/streamplot.py:212]: This can be outside the loop?

File: lib/matplotlib/streamplot.py [hunk 3]
```diff
@@ -193,26 +198,35 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
+            n = np.searchsorted(s, s[-1] * (x/(numarrows+1)))
```
Comment (jklymak) [inline lib/matplotlib/streamplot.py:213]: This can be vectorized outside the loop?

File: lib/matplotlib/streamplot.py [hunk 4]
```diff
@@ -193,26 +198,35 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
+            arrow_tail = (tx[n], ty[n])
+            arrow_head = (np.mean(tx[n:n + 2]), np.mean(ty[n:n + 2]))
+
+            if isinstance(linewidth, np.ndarray):
+                line_widths = interpgrid(linewidth, tgx, tgy)[:-1]
```
Comment (jklymak) [inline lib/matplotlib/streamplot.py:218]: this can be outside the loop

**Comment by jklymak** [comment]
This functionality seems reasonable.  

The code could be more efficient, moving things outside the loop.

Definitely needs a demo and a test proving it works.

File: lib/matplotlib/streamplot.py [hunk 5]
```diff
@@ -193,26 +199,37 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
         streamlines.extend(np.hstack([points[:-1], points[1:]]))
 
-        # Add arrows half way along each trajectory.
+        if use_multicolor_lines:
+            color_values = interpgrid(color, tgx, tgy)[:-1]
+            line_colors.append(color_values)
+
+        if numarrows <= -1:
+            raise ValueError("The value of numarrows must be"
+                             "greater than -1")
```
Comment (timhoffm) [inline lib/matplotlib/streamplot.py:208]:
```suggestion
        if numarrows < 0:
            raise ValueError("The value of numarrows must not"
                             "be negative")
```
Technically the same, but writing like this is more logical.

File: lib/matplotlib/streamplot.py [hunk 6]
```diff
@@ -56,6 +57,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     arrowstyle : str
         Arrow style specification.
         See `~matplotlib.patches.FancyArrowPatch`.
+    numarrows : int
+        Number of arrows per streamline.
+        The default is 1, which plots one arrow in the middle of the
+        streamline.
```
Comment (timhoffm) [inline lib/matplotlib/streamplot.py:63]:
```suggestion
    numarrows : int, default: 1
        Number of arrows per streamline. The arrows are
        distributed equidistantly on each line.
```
I think (did not follow all details of the code).

**Comment by clairefio** [comment]
Example output: 
numarrows=1
stream_9 updated

numarrows=3
stream_9

**Comment by jklymak** [comment]
@clairefio this could still use an example in our example gallery.  Thanks!

**Comment by clairefio** [comment]
> @clairefio this could still use an example in our example gallery. Thanks!

Ok, just added one!

**Comment by jklymak** [comment]
This PR is robustly failing tests, so some change you have made has affected other images.

**Comment by NilaiVemula** [comment]
I took a look at some of the build logs from Travis CI. There seem to be some problem that could be fixed:

- pyplot is not updated (not related to these changes): maybe rebase to master and update with `pip install -e`
- A lot of `ImageComparisonFailure` errors mostly to do with tests on stream plots: the new code should still produce the same output with the default parameters.

**Comment by dstansby** [comment]
Thanks for opening the PR @clairefio! Are you still planning on working on this? If so could you rebase this on to the current master branch? That way hopefully the tests should pass. Let me know if you have any questions/don't have time to pick this up at the moment.

**Pull request closed without merging**